codec_default.go

```diff
@@ -0,0 +1,48 @@
+package avro
+
+import (
+	"fmt"
+	"unsafe"
+
+	"github.com/modern-go/reflect2"
+)
+
+func createDefaultDecoder(cfg *frozenConfig, field *Field, typ reflect2.Type) ValDecoder {
+	fn := func(def any) ([]byte, error) {
+		defaultType := reflect2.TypeOf(def)
+		if defaultType == nil {
+			defaultType = reflect2.TypeOf((*null)(nil))
+		}
+		defaultEncoder := encoderOfType(cfg, field.Type(), defaultType)
+		if defaultType.LikePtr() {
+			defaultEncoder = &onePtrEncoder{defaultEncoder}
+		}
+		w := cfg.borrowWriter()
+		defaultEncoder.Encode(reflect2.PtrOf(def), w)
+		if w.Error != nil {
+			return nil, w.Error
+		}
+		return w.Buffer(), nil
+	}
+
+	b, err := field.encodeDefault(fn)
+	if err != nil {
+		return &errorDecoder{err: fmt.Errorf("decode default: %w", err)}
+	}
+	return &defaultDecoder{
+		defaultReader: cfg.borrowReader(b),
+		decoder:       decoderOfType(cfg, field.Type(), typ),
+	}
+}
+
+type defaultDecoder struct {
+	defaultReader *Reader
+	decoder       ValDecoder
+}
+
+// Decode implements ValDecoder.
+func (d *defaultDecoder) Decode(ptr unsafe.Pointer, _ *Reader) {
+	d.decoder.Decode(ptr, d.defaultReader)
```

Comment:
I think this is an issue. The reader will not reset on the second read, so will not read the default. I think the reader needs to move inside the Decode function, and the bytes should be passed through to the struct.
```suggestion
        r := cfg.borrowReader(d.data)
        defer cfg.returnReader(r)
        
	d.decoder.Decode(ptr, d.defaultReader)
```